src/core/border_router/routing_manager.cpp

```diff
@@ -2765,18 +2767,47 @@ void RoutingManager::RoutePublisher::ClearPreference(void)
 
 void RoutingManager::RoutePublisher::SetPreferenceBasedOnRole(void)
 {
-    UpdatePreference(Get<Mle::Mle>().IsRouterOrLeader() ? NetworkData::kRoutePreferenceMedium
-                                                        : NetworkData::kRoutePreferenceLow);
+    RoutePreference preference = NetworkData::kRoutePreferenceMedium;
+
+    if (Get<Mle::Mle>().IsChild() && (Get<Mle::Mle>().GetParent().GetTwoWayLinkQuality() != kLinkQuality1))
+    {
+        preference = NetworkData::kRoutePreferenceLow;
+    }
+
+    UpdatePreference(preference);
 }
 
-void RoutingManager::RoutePublisher::HandleRoleChanged(void)
+void RoutingManager::RoutePublisher::HandleNotifierEvents(Events aEvents)
 {
-    if (!mUserSetPreference)
+    VerifyOrExit(!mUserSetPreference);
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
     {
         SetPreferenceBasedOnRole();
     }
+
+    if (aEvents.Contains(kEventParentLinkQualityChanged))
+    {
+        VerifyOrExit(Get<Mle::Mle>().IsChild());
+
+        if (Get<Mle::Mle>().GetParent().GetTwoWayLinkQuality() == kLinkQuality1)
+        {
+            VerifyOrExit(!mTimer.IsRunning());
+            mTimer.Start(kDelayBeforePrfUpdateOnLinkQuality1);
+        }
+        else
+        {
+            UpdatePreference(NetworkData::kRoutePreferenceLow);
+            mTimer.Stop();
+        }
+    }
+
+exit:
+    return;
 }
 
+void RoutingManager::RoutePublisher::HandleTimer(void) { UpdatePreference(NetworkData::kRoutePreferenceMedium); }
```

Comment:
```suggestion
void RoutingManager::RoutePublisher::HandleTimer(void) { SetPreferenceBasedOnRole(); }
```

Ideally this would call `SetPreferenceBasedOnRole()` , because there could be code there (not now - but in the future) that determines the preference based on the current role. And current role still needs to be taken into account for the determination after the timer expires.

Reply:
Changed this in new push.